### PR TITLE
Add RWPI/ROPI relocation model support

### PR DESCRIPTION
**Pull request by alevy**
This PR adds support for using LLVM 4's ROPI and RWPI relocation models for ARM.

ROPI (Read-Only Position Independence) and RWPI (Read-Write Position Independence) are two new relocation models in LLVM for the ARM backend ([LLVM changset](https://reviews.llvm.org/rL278015)). The motivation is that these are the specific strategies we use in userspace [Tock](https://www.tockos.org) apps, so supporting this is an important step (perhaps the final step, but can't confirm yet) in enabling userspace Rust processes.

## Explanation

ROPI makes all code and immutable accesses PC relative, but not assumed to be overriden at runtime (so for example, jumps are always relative).

RWPI uses a base register (`r9`) that stores the addresses of the GOT in memory so the runtime (e.g. a kernel) only adjusts r9 tell running code where the GOT is.

## Complications adding support in Rust

While this landed in LLVM master back in August, the header files in `llvm-c` have not been updated yet to reflect it. Rust replicates that header file's version of the `LLVMRelocMode` enum as the Rust enum `llvm::RelocMode` and uses an implicit cast in the ffi to translate from Rust's notion of the relocation model to the LLVM library's notion.

My workaround for this currently is to replace the `LLVMRelocMode` argument to `LLVMTargetMachineRef` with an int and using the hardcoded int representation of the `RelocMode` enum. This is A Bad Idea(tm), but I think very nearly the right thing.

Would a better alternative be to patch rust-llvm to support these enum variants (also a fairly trivial change)?

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Hmm, this is a bit out of my wheelhouse. Perhaps @nagisa, @eddyb, or @michaelwoerister would be a better fit?

**Comment by nikomatsakis** [comment]
I didn't realize how short this PR is. @alevy, I think the main question is: is there any way to write a test for this?

**Comment by eddyb** [comment]
@alevy You should repeat what we did for other `LLVMRust` types, e.g. `LLVMRustCodeModel`.
That way it's really straight-forward to check that the names and values match up.

**Comment by alevy** [comment]
@eddyb thanks for the suggestion, I implemented something like `LLVMRelocCodeMode`, although due to compatibility issues with LLVM <= 3.8, there is a little bit of `ifdef` hell that seems required. I'm happy to change following any suggestions you have (particularly since I'm not sure I've captured the intent of the original `ifdef`s).

@nikomatsakis Tests would be nice, I'm just not sure where to put them. A quick grep for `LLVMRelocCodeMode` doesn't show up with any tests. Any pointers for where to place them?

File: src/rustllvm/PassWrapper.cpp [hunk 1]
```diff
@@ -239,6 +239,47 @@ static CodeGenOpt::Level fromRust(LLVMRustCodeGenOptLevel Level) {
...
 }
 
+enum class LLVMRustRelocMode {
+  LLVMRustRelocModeDefault,
+  LLVMRustRelocModeStatic,
+  LLVMRustRelocModePIC,
+  LLVMRustRelocModeDynamicNoPic,
+  LLVMRustRelocModeROPI,
+  LLVMRustRelocModeRWPI,
+  LLVMRustRelocModeROPIRWPI,
```
Comment (eddyb) [inline src/rustllvm/PassWrapper.cpp]: You don't need the prefix on each one, that's why it's `enum class`.

**Comment by eddyb** [comment]
@alevy I think the `ifdef`s look fine, there's mainly one style problem I mentioned.

File: src/librustc_llvm/ffi.rs [hunk 2]
```diff
@@ -288,6 +288,9 @@ pub enum RelocMode {
     Static = 1,
     PIC = 2,
     DynamicNoPic = 3,
+    ROPI = 4,
+    RWPI = 5,
+    ROPI_RWPI = 6,
```
Comment (eddyb) [inline src/librustc_llvm/ffi.rs]: If you won't have numbers in the C++ bindings, you should remove them here, btw.

File: src/rustllvm/PassWrapper.cpp (hunk 1 above)
Comment (alevy) [inline src/rustllvm/PassWrapper.cpp]: Oh duh.... fixed.

File: src/librustc_llvm/ffi.rs (hunk 2 above)
Comment (alevy) [inline src/librustc_llvm/ffi.rs]: OK, do you think it makes more sense to add numbers in the C++ or remove them here? (the other C++ enums don't have them, so I'm assuming remove them here)

Comment (eddyb) [inline src/librustc_llvm/ffi.rs]: Yeah, removing is fine. As long as they look the same, they'll behave the same.

**Comment by alevy** [comment]
OK @eddyb I think I've addressed your comments. Thanks!

**Pull request merged**